docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,246 @@
+
+#### Creating an Issue
+- Create the issue from the TA Projects Board (linked above)
+  - Click "add item" at the bottom of the `Todo` list
+  - Type `#autograder` to connect the issue to the Autograder
+  - Type a title and hit enter
+- Use a clear, descriptive title that uses one of these prefixes:
+  - `Backend:` server side changes
+  - `Frontend:` client (web) side changes
+  - `Fullstack:` for issues that require front and backend changes
+  - `Docs:` changes to documentation (either markdown files or JSDocs/JavaDocs)
+  - `Tests:` adding or modifying tests
+  - `Dev:` changes to enhance the Autograder development experience
+- Provide enough context so others can understand the purpose
+- Add labels as appropriate ([Here's the list](https://github.com/softwareconstruction240/autograder/issues/labels))
+- Link related issues
+- Issues should be narrow in scope
+
+#### Working on an Issue
+- Assign the issue to yourself before beginning work (this is how we know 
+  who is working on what, and gives you a sort of ownership over the issue)
+- Create a new branch ([more info here](#2-branches))
+- If you find a new issue, file it separately (don't mix concerns)
+- Leave a comment from time to time with updates on your progress
+
+### 2. Branches
+Branches in git allow us to all work on separate things in the same codebase.
+You will do all your work in a branch.
+
+#### Creating a branch
+Create the branch directly from the issue so your work is linked.
+
+- Always branch from `main` 
+  - (exceptions exist if you know what you're doing)
+- Use descriptive names that reflect what you're working on:
+    - `add-late-submission`
+    - `admin-display-verification-status`
+    - `server-communicator`
+    - `extract-service-logic-from-controllers`
+- Use kebab-case (lowercase with hyphens)
+- Keep names concise but clear
+- One branch per issue (don't mix different features/fixes)
+
+### 3. Making Changes
+Now that you have a branch, you can start to code! Hurray! 🎉
+Here are important guidelines to follow:
+
+- Review and follow the Style Guide (_Coming Soon_)
+- Write tests for your code. All new features and functionality must have tests demonstrating correctness
+  - Update existing tests if you change functionality
+  - _New contributors: A great way to better understand the codebase is to try writing tests for existing code_
+- Keep commits focused and meaningful
+  - Each commit should represent one logical change
+  - Write clear commit messages that describe that change ([See guidelines](#writing-good-commit-messages))
+- Update documentation as you go
+  - Add JSDoc/JavaDoc comments to new code
+    - _New contributors: Another great way to learn the codebase is to document existing code_
+  - Update repo Markdown docs if needed
+  - Code should for the most part be self-documenting. If not, make sure to add inline comments
+- Push your code to GitHub at the end of each coding session. This lets other
```

Comment:
It is a good policy to push all code at the end of a coding session, and is consistent with BYU's intellectual property policy to which we have all already agreed.

> ### Ownership
> Pursuant to law and university policy, **any work** (whether a Technical Work or a Creative Work) prepared by University Personnel within the scope of their employment, without an express agreement specifying otherwise, **is work for hire owned by the university**. 
> — [BYU Intellectual Property Policy](https://policy.byu.edu/view/intellectual-property-policy) (emphasis added)

Pushing the code into the BYU managed repo is one way we can give that IP to BYU. Conversely, keeping the work local is one way of keeping the IP from being in BYU's ownership.

Note that this policy doesn't specify that we have to push after "every commit," but only at the end of a coding session. This policy also doesn't restrict rebases. Rebasing a non-collaborative branch of any kind (feature or otherwise), requires a `git push --force`, but is otherwise safe the confident `git` user. This policy also doesn't preclude pushing the commits to other branches representing work-in-progress (WIP; this is another kind of branch besides feature branches).

This policy just requires that "the code is pushed," and that requirement is very consistent with the nature of our work in a managed environment for an employer.

Reply:
The IP Policy is a great reminder, I'm going to add that to the bottom of the document.

Reply:
I split the line, attempting to be more explicit in what we're expecting.